rasa/shared/utils/io.py

```diff
@@ -321,25 +322,106 @@ def env_var_constructor(loader, node):
     yaml.SafeConstructor.add_constructor("!env_var", env_var_constructor)
 
 
+class YAMLParser(yaml.YAML):
+    """A custom parser for YAML.
+
+    It's used because `ruamel.yaml` stores all its information globally.
+    Which means it's impossible to create different parsers with different constructors
+    and resolvers without remembering the previous configuration and restoring it once
+    it's needed. That's exactly what this class does.
+    """
+
+    def __init__(
+        self,
+        reader_type: Union[Text, List[Text]] = "safe",
+        replace_env_vars: bool = False,
+    ) -> None:
+        """Creates an instance of YAMLParser.
+
+        Args:
+            reader_type: YAML reader type.
+            replace_env_vars: Indicates if environment variables need to be replaced.
+        """
+        super().__init__(typ=reader_type)
+        self._save_default_yaml_parameters()
+
+        if replace_env_vars:
+            replace_environment_variables()
+        else:
+            # we need to do this because `yaml` adds contains a global `implicit_resolvers` list
+            # and the resolves from there cannot be de-registered
+            yaml.SafeConstructor.add_constructor("!env_var", lambda _, node: node.value)
+
+        self._save_modified_yaml_parameters()
+        self.restore_default_yaml_parameters()
+
+    def _save_default_yaml_parameters(self) -> None:
```

Comment:
Do we need a copy of the default saved in each parser object if they are always the same?